File: app/src/main/java/com/loopeer/codereader/ui/activity/MainActivity.java
package com.loopeer.codereader.ui.activity;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.ViewAnimator;

import com.loopeer.codereader.CodeReaderApplication;
import com.loopeer.codereader.Navigator;
import com.loopeer.codereader.R;
import com.loopeer.codereader.coreader.db.CoReaderDbHelper;
import com.loopeer.codereader.event.DownloadFailDeleteEvent;
import com.loopeer.codereader.model.Repo;
import com.loopeer.codereader.sync.DownloadRepoService;
import com.loopeer.codereader.ui.adapter.ItemTouchHelperCallback;
import com.loopeer.codereader.ui.adapter.MainLatestAdapter;
import com.loopeer.codereader.ui.decoration.DividerItemDecoration;
import com.loopeer.codereader.ui.decoration.DividerItemDecorationMainList;
import com.loopeer.codereader.ui.loader.ILoadHelper;
import com.loopeer.codereader.ui.loader.RecyclerLoader;
import com.loopeer.codereader.utils.RxBus;
import com.loopeer.directorychooser.FileNod;
import com.loopeer.directorychooser.NavigatorChooser;
import com.loopeer.itemtouchhelperextension.ItemTouchHelperExtension;

import java.util.List;

import butterknife.BindView;
import butterknife.OnClick;
import rx.android.schedulers.AndroidSchedulers;

public class MainActivity extends BaseActivity {
    private static final String TAG = "MainActivity";

    public static final int MY_PERMISSIONS_REQUEST_WRITE_EXTERNAL_STORAGE = 1000;
    @BindView(R.id.view_recycler)
    RecyclerView mRecyclerView;
    @BindView(R.id.animator_recycler_content)
    ViewAnimator mAnimatorRecyclerContent;
    @BindView(R.id.fab_main)
    FloatingActionButton mFabMain;

    private ILoadHelper mRecyclerLoader;
    private MainLatestAdapter mMainLatestAdapter;

    public ItemTouchHelperExtension mItemTouchHelper;
    public ItemTouchHelperExtension.Callback mCallback;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        getSupportActionBar().setDisplayHomeAsUpEnabled(false);
        Navigator.startDownloadRepoService(this, DownloadRepoService.DOWNLOAD_PROGRESS);
        if (ContextCompat.checkSelfPermission(this,
                Manifest.permission.WRITE_EXTERNAL_STORAGE)
                != PackageManager.PERMISSION_GRANTED) {
            if (ActivityCompat.shouldShowRequestPermissionRationale(this,
                    Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
            } else {
                ActivityCompat.requestPermissions(this,
                        new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE},
                        MY_PERMISSIONS_REQUEST_WRITE_EXTERNAL_STORAGE);
            }
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_repo_add, menu);
        getMenuInflater().inflate(R.menu.menu_settings, menu);
        getMenuInflater().inflate(R.menu.menu_github,menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_settings:
                Navigator.startSettingActivity(this);
                break;
            case R.id.action_repo_add:
                Navigator.startAddRepoActivity(this);
                break;
            case R.id.action_github:
                Navigator.startLoginActivity(this);
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onPostCreate(@Nullable Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        setUpView();
        registerSubscription(
                RxBus.getInstance().toObservable()
                        .filter(o -> o instanceof DownloadFailDeleteEvent)
                        .map(o -> ((DownloadFailDeleteEvent) o).deleteRepo)
                        .observeOn(AndroidSchedulers.mainThread())
                        .doOnNext(mMainLatestAdapter::deleteRepo)
                        .subscribe()
        );
    }

    @Override
    protected void onResume() {
        super.onResume();
        mRecyclerLoader.showProgress();
        loadLocalData();
    }

    private void setUpView() {
        mRecyclerLoader = new RecyclerLoader(mAnimatorRecyclerContent);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        mMainLatestAdapter = new MainLatestAdapter(this);
        mRecyclerView.setAdapter(mMainLatestAdapter);
        mRecyclerView.addItemDecoration(new DividerItemDecorationMainList(this,
                DividerItemDecoration.VERTICAL_LIST
                , getResources().getDimensionPixelSize(R.dimen.repo_list_divider_start)
                , -1
                , -1));
        mItemTouchHelper = createItemTouchHelper();
        mItemTouchHelper.attachToRecyclerView(mRecyclerView);
    }

    public ItemTouchHelperExtension createItemTouchHelper() {
        mCallback = createCallback();
        return new ItemTouchHelperExtension(mCallback);
    }

    public ItemTouchHelperExtension.Callback createCallback() {
        return new ItemTouchHelperCallback();
    }

    private void loadLocalData() {
        List<Repo> repos =
                CoReaderDbHelper.getInstance(CodeReaderApplication.getAppContext()).readRepos();
        setUpContent(repos);
    }

    @Override
    protected void reCreateRefresh() {
        super.reCreateRefresh();
        mRecyclerView.getRecycledViewPool().clear();
        mMainLatestAdapter.notifyDataSetChanged();
    }

    private void setUpContent(List<Repo> repos) {
        mRecyclerLoader.showContent();
        mMainLatestAdapter.updateData(repos);
    }

    @OnClick(R.id.fab_main)
    @SuppressWarnings("unused")
    public void onClick() {
        doSelectFile();
    }

    private void doSelectFile() {
        NavigatorChooser.startDirectoryFileChooserActivity(this);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        switch (requestCode) {
            case NavigatorChooser.DIRECTORY_FILE_SELECT_CODE:
                if (resultCode == RESULT_OK) {
                    FileNod node = (FileNod) data.getSerializableExtra(NavigatorChooser.EXTRA_FILE_NODE);
                    Repo repo = Repo.parse(node);
                    repo.id = String.valueOf(CoReaderDbHelper.getInstance(this).insertRepo(repo));
                    Navigator.startCodeReadActivity(MainActivity.this, repo);
                }
                break;
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        mMainLatestAdapter.clearSubscription();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           String permissions[], int[] grantResults) {
        switch (requestCode) {
            case MY_PERMISSIONS_REQUEST_WRITE_EXTERNAL_STORAGE: {
                if (grantResults.length > 0
                        && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                } else {
                }
                return;
            }
        }
    }

}

File: code-reader-kt/src/main/java/com/loopeer/codereaderkt/ui/activity/MainActivity.kt
package com.loopeer.codereaderkt.ui.activity

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ViewAnimator
import com.loopeer.codereaderkt.CodeReaderApplication
import com.loopeer.codereaderkt.Navigator
import com.loopeer.codereaderkt.R
import com.loopeer.codereaderkt.databinding.ActivityMainBinding
import com.loopeer.codereaderkt.db.CoReaderDbHelper
import com.loopeer.codereaderkt.event.DownloadFailDeleteEvent
import com.loopeer.codereaderkt.model.Repo
import com.loopeer.codereaderkt.sync.DownloadRepoService
import com.loopeer.codereaderkt.ui.adapter.ItemTouchHelperCallback
import com.loopeer.codereaderkt.ui.adapter.MainLatestAdapter
import com.loopeer.codereaderkt.ui.decoration.DividerItemDecoration
import com.loopeer.codereaderkt.ui.decoration.DividerItemDecorationMainList
import com.loopeer.codereaderkt.ui.loader.ILoadHelper
import com.loopeer.codereaderkt.ui.loader.RecyclerLoader
import com.loopeer.codereaderkt.utils.RxBus
import com.loopeer.directorychooser.FileNod
import com.loopeer.directorychooser.NavigatorChooser
import com.loopeer.itemtouchhelperextension.ItemTouchHelperExtension
import rx.android.schedulers.AndroidSchedulers


class MainActivity : BaseActivity() {

    private val TAG = "MainActivity"
    private val MY_PERMISSIONS_REQUEST_WRITE_EXTERNAL_STORAGE = 1000
    private lateinit var binding: ActivityMainBinding

    private lateinit var mRecyclerLoader: ILoadHelper
    private lateinit var mMainLatestAdapter: MainLatestAdapter

    lateinit var mItemTouchHelper: ItemTouchHelperExtension
    lateinit var mCallback: ItemTouchHelperExtension.Callback

    private lateinit var mRecyclerView: RecyclerView
    private lateinit var mAnimatorRecyclerContent: ViewAnimator
    @SuppressWarnings("unused")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)

        Navigator().startDownloadRepoService(this, DownloadRepoService.DOWNLOAD_PROGRESS)

        mRecyclerView = findViewById(R.id.view_recycler) as RecyclerView
        mAnimatorRecyclerContent = findViewById(R.id.animator_recycler_content) as ViewAnimator
        supportActionBar?.setDisplayHomeAsUpEnabled(false)
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            //获取权限
            if (ActivityCompat.shouldShowRequestPermissionRationale(this,
                    Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
            } else {
                ActivityCompat.requestPermissions(this,
                        arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
                        MY_PERMISSIONS_REQUEST_WRITE_EXTERNAL_STORAGE)
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_repo_add, menu)
        menuInflater.inflate(R.menu.menu_settings, menu)
        menuInflater.inflate(R.menu.menu_github, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_settings -> Navigator().startSettingActivity(this)
            R.id.action_repo_add -> Navigator().startAddRepoActivity(this)
            R.id.action_github -> Navigator().startLoginActivity(this)
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        setUpView()
        registerSubscription(
                RxBus.getInstance().toObservable()
                        .filter({ o -> o is DownloadFailDeleteEvent })
                        .map({ o -> (o as DownloadFailDeleteEvent).deleteRepo })
                        .observeOn(AndroidSchedulers.mainThread())
                        .doOnNext({ mMainLatestAdapter.deleteRepo(it) })
                        .subscribe()
        )
    }

    override fun onResume() {
        super.onResume()
        mRecyclerLoader.showProgress()
        loadLocalData()
    }

    private fun setUpView() {
        mRecyclerLoader = RecyclerLoader(mAnimatorRecyclerContent)
        mRecyclerView.layoutManager = LinearLayoutManager(this)
        mMainLatestAdapter = MainLatestAdapter(this)
        mRecyclerView.adapter = mMainLatestAdapter
        mRecyclerView.addItemDecoration(DividerItemDecorationMainList(this,
                DividerItemDecoration.VERTICAL_LIST, resources.getDimensionPixelSize(R.dimen.repo_list_divider_start), -1, -1))
        mItemTouchHelper = createItemTouchHelper()
        mItemTouchHelper.attachToRecyclerView(mRecyclerView)
    }

    private fun createItemTouchHelper(): ItemTouchHelperExtension {
        mCallback = createCallback()
        return ItemTouchHelperExtension(mCallback)
    }

    private fun createCallback(): ItemTouchHelperExtension.Callback = ItemTouchHelperCallback()

    private fun loadLocalData() {
        val repos = CoReaderDbHelper.getInstance(CodeReaderApplication.getAppContext()).readRepos()
        setUpContent(repos)
    }

    override fun reCreateRefresh() {
        super.reCreateRefresh()
        mRecyclerView.recycledViewPool.clear()
        mMainLatestAdapter.notifyDataSetChanged()
    }

    private fun setUpContent(repos: List<Repo>) {
        mRecyclerLoader.showContent()
        mMainLatestAdapter.updateData(repos)
    }

    fun onFabClick(view: View) {
        doSelectFile()
    }

    private fun doSelectFile() {
        NavigatorChooser.startDirectoryFileChooserActivity(this)
        //点击fab打开文件列表activity
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        super.onActivityResult(requestCode, resultCode, data)
        when (requestCode) {
            NavigatorChooser.DIRECTORY_FILE_SELECT_CODE -> if (resultCode == Activity.RESULT_OK) {
                val node = data.getSerializableExtra(NavigatorChooser.EXTRA_FILE_NODE) as FileNod
                val repo = Repo().parse(node)
                repo.id = CoReaderDbHelper.getInstance(this).insertRepo(repo).toString()
                Navigator().startCodeReadActivity(this@MainActivity, repo)
            }
        }
    }

    override fun onPause() {
        super.onPause()
        mMainLatestAdapter.clearSubscription()
    }

    override fun onRequestPermissionsResult(requestCode: Int,
                                            permissions: Array<String>, grantResults: IntArray) {
        when (requestCode) {
            MY_PERMISSIONS_REQUEST_WRITE_EXTERNAL_STORAGE -> {
                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                } else {
                }
                return
            }
        }
    }

}